### DISCUSSION TITLE
All-in-all very useful, but not so stable (refactoring request)

### DISCUSSION THREAD

**Discussion by simonferndriger** [alpinejs/alpine#3780] [4. General]
I am all-in-all very satisfied with what AlpineJS has to offer. However, there have been several encounters, where it just does not work properly. The side effects may be minor for each case (like prompting error messages because a template condition is captured too late, so the rendering already starts with an unavailable object), but in summary, they leave quite the impression that the whole framework is not stable. You just have to change a small thing somewhere, like for example adding a margin to a `x-intersect`, and all of a sudden the behaviour is totally unexpected and buggy.

For each case, it would take too long to analyse the issue deeply, extract the root problem and put it into a reproducable code snipped. So I just use what's working and delete or reverse what's not. This approach works for now, but I'm a but worried for the long run.

My wish: redo the whole AlpineJS from the ground, do a complete refactoring. Doing the same thing for the second time always makes it way better than it was before. And since AlpineJS is really great as a concept, it should be worth the time.

**Comment by trych** [comment]
"I have these vague issues with Alpine that I am not willing to reproduce properly, so please rewrite the entire thing."

lmao ... what?

P.S.: While I also encountered some issues with `x-intersect` once that I was too lazy to debug at the time, this is not even part of the Alpine core. Alpine itself seems very stable to me, generally.

**Comment by ekwoka** [comment]
I've worked in Alpine a lot (I'm basically just a profession Alpine dev), and seen every issue people come here with.

Alpine isn't perfect but 99.9% of the time it's been PEBKAC when these errors arise. The more advanced issues have generally been in how browsers behave for certain actions, which isn't related to Alpine, just the thing I was doing with it.

I am on the side working on a full refactor/recreation to increase performance and stability though.

The reactivity engine is already much faster and smaller.

Some things are ported to Alpine as I go.

I'd be curious to see more examples of these "condition captured late" or whatever.

**Reply by simonferndriger** [reply]
I'm sure that it is not easy considering all browser and application based difficulties. So, I really appreciate your effort! 🙏

About your questions. As of now, I would say the the most hindering issues are:
1. https://github.com/alpinejs/alpine/discussions/3693
2. https://github.com/alpinejs/alpine/discussions/3552 (would be nice if the workaround would not be necessary) 

And sometimes I just get a module error without knowing the cause, in this case for example I suspect it's because of a for loop:
```
Uncaught (in promise) TypeError: i is not a function
    at module.esm.js:1068:28
```

**Reply by ekwoka** [reply]
Yeah, unfortunately on the last one, it's annoying that so many people that use alpine use 16 year old build systems so Caleb reverted the thing that allowed directly importing the source to get better feedback in dev mode.

I'll look more into both of those first bugs. The second is one I've seen as an issue before. Template cloned elements are not cleaned up at the same time as the template itself is cleaned up, which should probably happen inline.

**Reply by simonferndriger** [reply]
Thank you!

**Reply by simonferndriger** [reply]
PS: I'm using Vite now, and am quite happy with it 😊

**Reply by simonferndriger** [reply]
A third one has probably something to do with a `<template x-for>` being rendered (half-way?) even though the looped element is empty. Because I often get the error `undefined is not an object (evaluating 'poi(p).info')` in `sentry.io` for the following code:

![image](https://github.com/alpinejs/alpine/assets/1435963/7665b76a-07e2-4c7f-b3e2-57da585f6c2c)

**Reply by ekwoka** [reply]
That information is confusing, and I don't know the structures of the things, but on look it seems wrong.

This is likely because you don't have a :key set on the template so making changes to the list causes issues.

**Reply by simonferndriger** [reply]
OK, I'll try that with the key - sorry, forgot about that

**Reply by simonferndriger** [reply]
Using a key did not solve the issue, but I'm just using `poi(p)?.` now to avoid all the console errors

**Reply by ekwoka** [reply]
why are you even calling a function like that to begin with? that may be the real cause.

That `poi(p)` starts returning null while `p` is still rendered in the loop

Why not do like

```js
x-for="p in pois.map(poi)"
```

Then p can be the data, and you won't need to keep calling that function (idk what it does but it may be costly?)

and presumably the key was not just index, right?

**Reply by simonferndriger** [reply]
Ah OK, I begin to understand. I think you might be right about your assumption.

Well, I need this poi() function because pois contains the raw data from different sources, and poi() does standardize it. But you're right, it probably makes more sense to initially map it, so I tried that and it seems to solve the issue! 😀

`initPois(data = []){
        this.pois = data.map(p => this.poi(p)) || [];
    },`

**Reply by simonferndriger** [reply]
PEBKAC++

**Reply by simonferndriger** [reply]
PS: I changed/corrected #2 (looks like there is no issue with two template siblings)

**Reply by simonferndriger** [reply]
There is still an issue. Now I know when it occurs. It's when the pois are still rendering and I'm changing the array the is being loop through. The console output is the following:

![image](https://github.com/alpinejs/alpine/assets/1435963/b2bcad53-5ebd-476c-9ead-f28fb50d0cf9)

While the code looks like this:

![image](https://github.com/alpinejs/alpine/assets/1435963/076e6d97-fe03-4815-a0be-808837fc8551)

**Reply by simonferndriger** [reply]
So I guess the for loop is still rendering while the poi array gets first set to an empty array before the new fetch is completed. To avoid this issue I use this workaround, but it would be nice if the for loop would check this by itself and only render items that are currently in the array:

![image](https://github.com/alpinejs/alpine/assets/1435963/55f58b84-f677-4167-8468-5833e6896b01)

**Reply by SimoTod** [reply]
Yeah, your array has elements. They are undefined but that's not something Alpine checks for you, the array could also contains objects that don't have the property you expect and those objects would trigger the same error.
Do you need to initialise the array that way if you then populate it? 
Bear in mind that when you change the array, you force Alpine to recompute the for loop so it's better to change it as little as possible. If you really have to, there's a way to pause the reactivity.

**Reply by ekwoka** [reply]
Wyou can just initialize it as an empty array, instead of an an array of undefined elements...

**Reply by simonferndriger** [reply]
@SimoTod Ah, interesting. How can you pause the reactivity (in general or just for this specific loop?

**Reply by simonferndriger** [reply]
@ekwoka Actually, I was assuming that AlpineJS was creating this array of undefined objects, no? Because I can't imagine how this can happen with my code since I do set it to an empty array in between changes; although it's possible I guess  - but how?

This is the part that refreshes the pois:
![image](https://github.com/alpinejs/alpine/assets/1435963/d436db79-54c4-4895-95f4-5bf712c6e29c)

Reset is clearing the array:
![image](https://github.com/alpinejs/alpine/assets/1435963/ba1a0556-79b6-4355-84e9-4e939bc41425)

And the new data gets assigned here (it happened for the hotels):
![image](https://github.com/alpinejs/alpine/assets/1435963/31096736-30d5-44c3-8ac5-d8df8ed6be54)

While initPois() as a last step even has a fallback to an empty array if something went wrong before:
![image](https://github.com/alpinejs/alpine/assets/1435963/a34a7383-cd26-44ac-904e-1599c283b6d7)

At last, even the poi() mapping has an object as fallback:
![image](https://github.com/alpinejs/alpine/assets/1435963/396dc6af-5109-42b1-8e61-0176fda5039b)

**Reply by ekwoka** [reply]
The `data.map || []` will never trigger, btw.


That is confusing.

Definitely shouldn't be happening but I can't quite tell in this why that would be happening...

I'd suggest changing all the .then callbacks to async await and then slapping a `debugger;` at the start of load and follow it through...

Where do you initialize `pois`?

**Reply by SimoTod** [reply]
@simonferndriger There are a few methods on the Alpine object but I don't think they are documented because they are mostly internals: deferMutations, flushAndStopDeferringMutations, startObservingMutations, and stopObservingMutations.
Can you post your `poi` function?

**Reply by simonferndriger** [reply]
@ekwoka 😂 You are right, since `map()` always returns an Array my addition can be left out, thanks for the hint 😊

**Reply by simonferndriger** [reply]
@SimoTod 

Sure, here is my complete `poi()`:

![image](https://github.com/alpinejs/alpine/assets/1435963/764e8be5-0d48-433d-bb6a-b35e229bfd28)

**Reply by ekwoka** [reply]
To me, I'd see this as there being a case that is not handled. and so the function isn't returning.

Also, since I see you're using cloudinary, I do actually [have an alpine plugin for that](https://www.npmjs.com/package/@ekwoka/x-rias) 😄  Maybe it doesn't support your specific square fill option (yet, it could), but it can generate srcset



Oh, you trigger load, and in between the fetch and and the initPois being called, webcam changes. that would give this result you are experiencing.

**Reply by simonferndriger** [reply]
@ekwoka Ah, you are right! I totally forgot about the case where there is no active element but the loading still in the works. Therefore this solved the issue:

![image](https://github.com/alpinejs/alpine/assets/1435963/e1da81f6-ad12-44e2-81fd-c6f727789be3)


Thanks for the hint about your cloudinary srcset plugin, but I think in my case, a single size is sufficient for now 😊

**Reply by ekwoka** [reply]
I'd probably do this check before the `this.pois =`

like

```js
this.pois = this.active ? data.map(p => this.poi(p)) : []
```

Otherwise, you'll still end up with a bunch of empty elements doing work for nothing

**Reply by simonferndriger** [reply]
Top! You're right, that's even beter, thanks 😊

**Reply by simonferndriger** [reply]
Very sorry, but there seems still to be a small issue. How can I solve this? `pois` seems to be undefined sometimes, even though I don't know why, at least with `(pois || [])` this should never be an issue, but obviously I'm wrong...?

![image](https://github.com/alpinejs/alpine/assets/1435963/bcb9d7b9-47a0-422f-87fb-fb1eb5a63522)

**Reply by ekwoka** [reply]
That is quite strange for sure....

**Reply by simonferndriger** [reply]
Is it possible, that it actually has something to do with the x-for implementation?

**Reply by ekwoka** [reply]
if `pois` is a function, and not an array, this could happen.

**Reply by simonferndriger** [reply]
`pois` is never a function, but it's also the name of the Alpine data container. Could this be a problem?

![image](https://github.com/alpinejs/alpine/assets/1435963/4d9c34c5-277c-4c84-aa33-600a3be09660)

![image](https://github.com/alpinejs/alpine/assets/1435963/1a6eb209-acab-4aad-afd3-c0dc05102ea5)

**Reply by simonferndriger** [reply]
@ekwoka Is there maybe a quickfix for this such that at least the error message can be prevented?

**Reply by ekwoka** [reply]
Not just off what I'm seeing there.

Is it destructive? like it actually doesn't correct itself? or just an error in a branch?

Could be related to just changing the values two quickly and the error is not in evaluating `pois` but in evaluating something else in the x-for

**Reply by simonferndriger** [reply]
Since everything seems to work, I guess it's "just" an error...

**Comment by simonferndriger** [comment]
OK, I actually got an urgent matter. There is this `<section>` which has a fix position with `overflow:auto`. When I use the condition to show it via `<template>` it works fine, but as soon as I use the very same condition with `x-show`, the condition fails to evaluate correctly as soon as I start scrolling the overlay. I checked with `$watch`, none of the condition parameters changes while scrolling, just the evaluation result.

This is especially problematic, because I have to load ads into this area - and I just found out, that I need to load them instantly, when loading them delayed via `<template>` tag, they don't show up (without actually any error) - ads are just the worst code ever, but you have to life with it...

**Reply by ekwoka** [reply]
It's have to see the code

**Reply by simonferndriger** [reply]
Sure. These are the steps to reproduce (on the project):

1. Visit https://rc.swisswebcams.ch/?design=new with a big enough desktop that you see 2 columns below the intro
2. Click on any webcam
3. If you scroll now the right side while the mouse is over it (you might have to shrink your window if you have a big screen), eveything works as expected because the `<template "(!webcam && columns === 2) || ad.show">` wrapper is currently in use for the ad section
4. When you remove the template and add `x-show="(!webcam && columns === 2) || ad.show"` to the `<section x-data="player('ad')" ...>` directly and scroll around a bit in the webcam area on the right side, you'll see that the ad section is also showing; and there is no "x-show" defined in "player", such that it could get overridden, I checked for that.

![image](https://github.com/alpinejs/alpine/assets/1435963/1270edba-23d1-451b-a3ef-65b9feadba33)

![image](https://github.com/alpinejs/alpine/assets/1435963/4fc6ab36-842a-4e81-bd60-021ccb5b0d93)

**Reply by simonferndriger** [reply]
Ah, I see now that the problem was `[':style']`, this must have overridden the `x-show` in the tag itself. Not ideal in my opinion, because it can easily be overlooked, but I can understand the logic behind it.

**Reply by ekwoka** [reply]
Yes, if you use a string syntax for style, it will conflict. You should be able to use object syntax for style and it should work, or even put the display attribute directly into style.

**Reply by simonferndriger** [reply]
Ah, of course, I should have been using the object syntax - didn't think of that. But this way it works as expected, great, thx!

![image](https://github.com/alpinejs/alpine/assets/1435963/3cf722d5-2f06-4247-b438-2e15e2a03a86)

**Reply by simonferndriger** [reply]
It's just with classes, that the string syntax is additive then?

**Reply by ekwoka** [reply]
Yes, because DOMTokenList provides methods to be able to do that. style does not, it's just a string or manual assignment of styles from the object.

So it's a limitation in alpine that stems from a limitation in the DOM APIs. There isn't a good way to parse and combine string styles.

**Reply by simonferndriger** [reply]
OK, I understand.
